### PR TITLE
Integrate Universal SRS from Setup

### PR DESCRIPTION
**Pull request by Pratyush**
## Motivation

Integrate the Universal SRS from the Setup into our system
## Test Plan

(Write your test plan here)

## Related PRs

(Link your related PRs here)

### CODE REVIEW

File: algorithms/src/polycommit/kzg10/data_structures.rs [hunk 1]
```diff
@@ -26,128 +26,81 @@ use snarkvm_utilities::{
...
 use anyhow::Result;
 use core::ops::{Add, AddAssign};
-use parking_lot::RwLock;
 use rand_core::RngCore;
-use std::{collections::BTreeMap, io, sync::Arc};
+use std::{collections::BTreeMap, io, ops::Range};
 
 /// `UniversalParams` are the universal parameters for the KZG10 scheme.
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, CanonicalSerialize, CanonicalDeserialize)]
```
Comment (howardwu) [inline algorithms/src/polycommit/kzg10/data_structures.rs]: What size of `PowersOfG` does Canonical* write to bytes? And what does it expect when reading in?

File: algorithms/src/polycommit/kzg10/data_structures.rs [hunk 2]
```diff
@@ -232,19 +154,7 @@ impl<E: PairingEngine> ToBytes for UniversalParams<E> {
...
-
-    pub fn supported_degree_bounds(&self) -> &[usize] {
-        &self.supported_degree_bounds
-    }
-
-    pub fn increase_degree(&self, degree: usize) -> Result<()> {
-        self.download_up_to(degree)
-    }
-}
+impl<E: PairingEngine> UniversalParams<E> {}
```
Comment (howardwu) [inline algorithms/src/polycommit/kzg10/data_structures.rs]: Is this still needed?

File: algorithms/src/polycommit/kzg10/mod.rs [hunk 3]
```diff
@@ -96,121 +93,11 @@ pub struct KZG10<E: PairingEngine>(PhantomData<E>);
...
-            inverse_neg_powers_of_beta_h,
-            prepared_h,
-            prepared_beta_h,
-        };
-        end_timer!(setup_time);
-        Ok(pp)
+    pub fn load_srs(max_degree: usize) -> Result<UniversalParams<E>, PCError> {
+        let params = UniversalParams::load()?;
+        params.download_powers_for(0..(max_degree + 1))?;
+        // TODO: check that params match supported_degree bounds.
```
Comment (howardwu) [inline algorithms/src/polycommit/kzg10/mod.rs]: Is this TODO needed?

File: algorithms/src/polycommit/kzg10/data_structures.rs (hunk 1 above)
Comment (Pratyush) [reply algorithms/src/polycommit/kzg10/data_structures.rs]: Whatever is in memory (it writes out the size of the existing powers)

File: algorithms/src/polycommit/test_templates.rs [hunk 4]
```diff
@@ -283,8 +283,7 @@ where
                 None
             };
 
-            let hiding_bound =
-                if num_points_in_query_set >= degree { Some(degree) } else { Some(num_points_in_query_set) };
+            let hiding_bound = Some(1);
```
Comment (howardwu) [inline algorithms/src/polycommit/test_templates.rs:286]: Only to confirm: is this the right value to check in this test?

Comment (Pratyush) [reply algorithms/src/polycommit/test_templates.rs:286]: Yes; we actually do not have the extra powers for satisfying the old hiding_bound, so we have to change the hiding_bound to `Some(1)`

File: algorithms/src/polycommit/kzg10/mod.rs (hunk 3 above)
Comment (Pratyush) [reply algorithms/src/polycommit/kzg10/mod.rs]: No, removed in [`1e21f75` (#1245)](https://github.com/AleoHQ/snarkVM/pull/1245/commits/1e21f7525c2d54d68146fada724bc611092b75f7)

File: algorithms/src/polycommit/kzg10/data_structures.rs (hunk 2 above)
Comment (Pratyush) [reply algorithms/src/polycommit/kzg10/data_structures.rs]: No, removed in [`1e21f75` (#1245)](https://github.com/AleoHQ/snarkVM/pull/1245/commits/1e21f7525c2d54d68146fada724bc611092b75f7)

File: parameters/src/testnet3/powers.rs [hunk 5]
```diff
@@ -256,112 +202,289 @@ impl<E: PairingEngine> ToBytes for PowersOfG<E> {
...
+                final_powers = additional_powers;
+            } else {
+                final_powers.extend(additional_powers);
             }
-            // Return the powers.
-            Ok(Self::setup(powers_of_beta_g, powers_of_beta_times_gamma_g)?)
-        } else {
-            Ok(Self::load()?)
         }
+        final_powers.extend(self.shifted_powers_of_beta_g.drain(..self.shifted_powers_of_beta_g.len()));
```
Comment (howardwu) [inline parameters/src/testnet3/powers.rs]:
```suggestion
        final_powers.extend(self.shifted_powers_of_beta_g.iter());
```

File: parameters/src/testnet3/powers.rs [hunk 6]
```diff
@@ -256,112 +202,289 @@ impl<E: PairingEngine> ToBytes for PowersOfG<E> {
...
+
+        let mut download_queue = Vec::with_capacity(14);
+        let mut existing_num_powers = self.shifted_powers_of_beta_g.len();
+        while existing_num_powers < extra_powers_to_download {
+            existing_num_powers *= 2;
+            download_queue.push(existing_num_powers);
+        }
+        download_queue.reverse(); // We want to download starting from the smallest power.
+
+        let mut final_powers = vec![];
```
Comment (howardwu) [inline parameters/src/testnet3/powers.rs]:
```suggestion
        let mut final_powers = Vec::with_capacity(extra_powers_to_download);
```

File: parameters/src/testnet3/powers.rs [hunk 7]
```diff
@@ -256,112 +202,289 @@ impl<E: PairingEngine> ToBytes for PowersOfG<E> {
...
+    pub fn download_powers_for(&mut self, range: &Range<usize>) -> Result<()> {
+        if self.contains_in_normal_powers(range) || self.contains_in_shifted_powers(range) {
+            return Ok(());
+        }
+        let half_max = MAX_NUM_POWERS / 2;
+        if (range.start <= half_max) && (range.end > half_max) {
+            // If the range contains the midpoint, then we must download all the powers.
+            // (because we round up to the next power of two).
+            self.download_up_to(range.end)?;
+            self.shifted_powers_of_beta_g = Vec::new();
```
Comment (howardwu) [inline parameters/src/testnet3/powers.rs:321]:
```suggestion
            self.shifted_powers_of_beta_g.clear();
```

File: parameters/src/testnet3/powers.rs (hunk 6 above)
Comment (howardwu) [reply parameters/src/testnet3/powers.rs]: Added in recent commit

File: parameters/src/testnet3/powers.rs (hunk 5 above)
Comment (howardwu) [reply parameters/src/testnet3/powers.rs]: Added in recent commit

**Review by howardwu — APPROVED** [review]
LGTM

File: parameters/src/testnet3/powers.rs (hunk 7 above)
Comment (Pratyush) [reply parameters/src/testnet3/powers.rs:321]: Doesn't `clear` keeps the old capacity? W e will never use that allocated memory again, which is why used `new` to free that memory ASAP.

Comment (howardwu) [reply parameters/src/testnet3/powers.rs:321]: Yes, will revert

**Review by howardwu — APPROVED** [review]
LGTM

**Pull request merged**